File: pkg/pillar/types/monitorapi/tpm.go
// Copyright (c) 2026 Zededa, Inc.
// SPDX-License-Identifier: Apache-2.0

package monitorapi

// EFIVariable is a single UEFI variable captured alongside a measured boot,
// used by the TUI to explain PCR mismatches. Value is raw bytes (base64 on the
// wire).
type EFIVariable struct {
	Name  string `json:"name,omitempty"`
	Value []byte `json:"value,omitempty"`
}

// TpmLogs carries the raw TPM measured-boot event logs and the UEFI boot
// variables for the last good and last failed boots. The TUI diffs these
// against the locked vault's PCRs to show what changed. The log payloads are
// opaque binary blobs (base64 on the wire); the TUI parses them client-side.
type TpmLogs struct {
	LastFailedLog   []byte        `json:"last_failed_log,omitempty"`
	LastGoodLog     []byte        `json:"last_good_log,omitempty"`
	BackupFailedLog []byte        `json:"backup_failed_log,omitempty"`
	BackupGoodLog   []byte        `json:"backup_good_log,omitempty"`
	EFIVarsSuccess  []EFIVariable `json:"efi_vars_success,omitempty"`
	EFIVarsFailed   []EFIVariable `json:"efi_vars_failed,omitempty"`
}

File: pkg/monitor/src/ipc/tpm.rs
// Copyright (c) 2026 Zededa, Inc.
// SPDX-License-Identifier: Apache-2.0

//! Client-side helpers on the generated `TpmLogs` wire type.

use super::monitorapi::TpmLogs;
use anyhow::Result;
use std::fs::File;
use std::io::Write;

impl TpmLogs {
    /// Dump the raw binary measured-boot logs to `path` for offline analysis.
    /// Empty (absent) logs are skipped.
    pub fn save_raw_binary_logs(&self, path: &str) -> Result<()> {
        for (name, data) in [
            ("last_failed_log", &self.last_failed_log),
            ("last_good_log", &self.last_good_log),
            ("backup_failed_log", &self.backup_failed_log),
            ("backup_good_log", &self.backup_good_log),
        ] {
            if !data.is_empty() {
                File::create(format!("{path}/{name}.bin"))?.write_all(data)?;
            }
        }
        Ok(())
    }
}
